### PR TITLE
feat: add robot actions

### PR DESCRIPTION
**Pull request by dido18**
## Motivation 
Add scratch blocks for the Arduino Robot extension.

## Changes
Added a `Robot.cpp` class in the sketch 
Added the following scratch blocks:
- move forward for X steps
- move backward fox X steps
-  move forward for X seconds
- move backward fox X seconds
- turn right for X seconds
- turn left for X seconds
- set speed to [0-100] percentage

New scratch blocks

### CODE REVIEW

File: sketches/BLE_Scratch/BLE_Scratch.ino [hunk 1]
```diff
@@ -409,13 +425,13 @@ void onPinActionCharacteristicWrite(BLEDevice central, BLECharacteristic charact
...
-        }
-        n = n->next;
-      }
+       n = root;
+       while (n != NULL) {
+         if (n->pin == pinNumber) {
+           break;
+         }
+         n = n->next;
+       }
```
Comment (lucarin91) [inline sketches/BLE_Scratch/BLE_Scratch.ino:434]: can you revert those changes?

File: sketches/BLE_Scratch/BLE_Scratch.ino [hunk 2]
```diff
@@ -495,3 +511,36 @@ void setLedPinValue(int pin, int value) {
...
 }
+
+void onRobotActionCharacteristicWrite(BLEDevice central, BLECharacteristic characteristic) {
+  enum pinAction action = (enum pinAction)pinRobotCharacteristic[0];
+  int8_t steps = pinActionCharacteristic[1];
+
+  Serial.print("ROBOT action=");
+  Serial.println(action);
+  Serial.print("steps = ");
+  Serial.println(steps);
```
Comment (lucarin91) [inline sketches/BLE_Scratch/BLE_Scratch.ino:522]: here you should do the print only if the `Serial` is enabled
```suggestion
  if (Serial){
      Serial.print("ROBOT action=");
      Serial.println(action);
      Serial.print("steps = ");
      Serial.println(steps);
  }
```

File: sketches/BLE_Scratch/BLE_Scratch.ino [hunk 3]
```diff
@@ -495,3 +511,36 @@ void setLedPinValue(int pin, int value) {
...
+  switch (action) {
+    case robotAction::MOVE_FORWARD:
+      Serial.println("go forward");
```
Comment (lucarin91) [inline sketches/BLE_Scratch/BLE_Scratch.ino:525]: I will remove this print line

File: sketches/BLE_Scratch/BLE_Scratch.ino [hunk 4]
```diff
@@ -495,3 +511,36 @@ void setLedPinValue(int pin, int value) {
...
+      myra.moveForward(steps);
+      break;
+    case robotAction::MOVE_BACKWARD:
+      Serial.println("go backward");
+      myra.moveBackward(steps);
+      break;
+    case robotAction::TURN_LEFT:
+      Serial.println("turn left");
```
Comment (lucarin91) [inline sketches/BLE_Scratch/BLE_Scratch.ino:533]:
```suggestion
```

File: sketches/BLE_Scratch/BLE_Scratch.ino [hunk 5]
```diff
@@ -495,3 +511,36 @@ void setLedPinValue(int pin, int value) {
...
+  Serial.println(action);
+  Serial.print("steps = ");
+  Serial.println(steps);
+  switch (action) {
+    case robotAction::MOVE_FORWARD:
+      Serial.println("go forward");
+      myra.moveForward(steps);
+      break;
+    case robotAction::MOVE_BACKWARD:
+      Serial.println("go backward");
```
Comment (lucarin91) [inline sketches/BLE_Scratch/BLE_Scratch.ino:529]: also this
```suggestion
```

File: sketches/BLE_Scratch/BLE_Scratch.ino [hunk 6]
```diff
@@ -495,3 +511,36 @@ void setLedPinValue(int pin, int value) {
...
+      myra.moveBackward(steps);
+      break;
+    case robotAction::TURN_LEFT:
+      Serial.println("turn left");
+      myra.turnLeft();
+      break;
+    case robotAction::TURN_RIGHT:
+      Serial.println("turn right");
```
Comment (lucarin91) [inline sketches/BLE_Scratch/BLE_Scratch.ino:537]: also this
```suggestion
```

File: sketches/BLE_Scratch/BLE_Scratch.ino [hunk 7]
```diff
@@ -495,3 +511,36 @@ void setLedPinValue(int pin, int value) {
...
+      myra.turnRight();
+      break;
+    case robotAction::STOP:
+      Serial.println("stop");
```
Comment (lucarin91) [inline sketches/BLE_Scratch/BLE_Scratch.ino:541]:
```suggestion
```

File: sketches/BLE_Scratch/robot.cpp [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+Robot::Robot(int rightWheelPin, int leftWheelPin) {
+  rightWheel = Servo();
+  leftWheel = Servo();
+  rightWheel.attach(rightWheelPin);
+  leftWheel.attach(leftWheelPin);
+  speed = 10;
+  // used to calibrate the middle point 
+  rightWheel.write(90);
+  leftWheel.write(90);
+}
```
Comment (lucarin91) [inline sketches/BLE_Scratch/robot.cpp:13]: I usually see the constructor in the `.h`, I don't know which are the implication. I also prefer to use `this` to explicitly show which are the class member.
```suggestion
Robot::Robot(int rightWheelPin, int leftWheelPin) {
  this->rightWheel = Servo();
  this->leftWheel = Servo();
  this->rightWheel.attach(rightWheelPin);
  this->leftWheel.attach(leftWheelPin);
  this->speed = 10;
  // used to calibrate the middle point 
  this->rightWheel.write(90);
  this->leftWheel.write(90);
}
```

File: sketches/BLE_Scratch/robot.cpp [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+void Robot::turnRight() {
+    rightWheel.write(90 + speed);
+    leftWheel.write(90 + speed);
+    delay(700);
+    rightWheel.write(90);
+    leftWheel.write(90);
+}
+
+void Robot::turnLeft() {
+    rightWheel.write(90 - speed );
```
Comment (lucarin91) [inline sketches/BLE_Scratch/robot.cpp:45]:
```suggestion
    rightWheel.write(90 - speed);
```

File: sketches/BLE_Scratch/robot.cpp [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+void Robot::moveBackward(int steps) {
+  // TODO: use millis() instead of delay
+  // for loop on steps
+  rightWheel.write(90 - speed);
+  leftWheel.write(90 + speed);
+  delay(1000);
+  rightWheel.write(90);
+  leftWheel.write(90);
```
Comment (lucarin91) [inline sketches/BLE_Scratch/robot.cpp:32]: probably the best way to stop the servos is to detach them, in the same way it is done in the stop method of the servo.

File: sketches/BLE_Scratch/robot.h [hunk 11]
```diff
@@ -0,0 +1,26 @@
+#ifndef Robot_h
+#define Robot_h
+
+#include <Arduino.h>
+#include <Servo.h>
+
+class Robot {
```
Comment (lucarin91) [inline sketches/BLE_Scratch/robot.h:7]: I usually prefer to use `struct` instead, AFAIK the only difference between class and struct is that `struct` has all member public by default, instead the classes have all members private by default.
```suggestion
struct Robot {
```

File: sketches/BLE_Scratch/robot.h [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+  Servo rightWheel;
+  Servo leftWheel;
+  int speed;
+  /*
+      [0,89]     => go forward (from faster to slower)
+      90         => stop
+      [91-180]   => go backward (from lower to faster)
+     */
+
+public:
```
Comment (lucarin91) [inline sketches/BLE_Scratch/robot.h:18]: If you use struct you can remove this 
```suggestion
```

**Review by lucarin91 — COMMENTED** [review]
Good job 👍 I added some minors comments

File: sketches/BLE_Scratch/robot.h (hunk 11 above)
Comment (dido18) [reply sketches/BLE_Scratch/robot.h:6]: Let's keep `class`

File: scratch-arduino-extensions/scratch-vm/src/extensions/scratch3_arduino/peripheral.js [hunk 13]
```diff
@@ -476,6 +484,43 @@ class ArduinoPeripheral {
         return this._send(BLECommand.PIN_ACTIONS, pinAction);
     }
 
+    moveForward(steps) {
+        const robotAction = [
+            ROBOT_ACTIONS.MOVE_FORWARD,
+            steps,
+            0xFF
```
[the remark this answers is not in the document]
Comment (dido18) [reply scratch-arduino-extensions/scratch-vm/src/extensions/scratch3_arduino/peripheral.js]: Removed.

File: sketches/BLE_Scratch/BLE_Scratch.ino [hunk 14]
```diff
@@ -4,14 +4,12 @@
...
+// #define ARDUINO_NANO_BLE_SENSE
 //#define ARDUINO_NANO_BLE
-// #define ARDUINO_NANO_BLE_SENSE_R2
+ #define ARDUINO_NANO_BLE_SENSE_R2
 
 // when this is defined the board will print debug messages on serial
-//#define DEBUG
-
-
+#define DEBUG
```
Comment (lucarin91) [inline sketches/BLE_Scratch/BLE_Scratch.ino:12]: I will keep this disabled by default
```suggestion
// #define DEBUG
```

File: sketches/BLE_Scratch/BLE_Scratch.ino [hunk 15]
```diff
@@ -111,13 +117,11 @@ if (!APDS.begin()) {
 // All 3 variants have an IMU on it
   if (!IMU.begin()) {
     printSerialMsg("Failled to initialized IMU!");
-
```
Comment (lucarin91) [inline sketches/BLE_Scratch/BLE_Scratch.ino:114]: can you revert this?

File: sketches/BLE_Scratch/BLE_Scratch.ino [hunk 16]
```diff
@@ -111,13 +117,11 @@ if (!APDS.begin()) {
...
     while (1);
   }
 
   if (!BLE.begin()) {
     printSerialMsg("Failled to initialized BLE!");
-
```
Comment (lucarin91) [inline sketches/BLE_Scratch/BLE_Scratch.ino:120]: can you revert this?

File: sketches/BLE_Scratch/BLE_Scratch.ino [hunk 17]
```diff
@@ -495,3 +505,50 @@ void setLedPinValue(int pin, int value) {
...
+  MOVE_BACKWARD_STEP = 1,
+  TURN_RIGHT = 2,
+  TURN_LEFT = 3,
+  SET_SPEED = 4,
+  MOVE_FORWARD_TIME = 5,
+  MOVE_BACKWARD_TIME = 6,
+};
+
+void onRobotActionCharacteristicWrite(BLEDevice central, BLECharacteristic characteristic) {
+  enum pinAction action = (enum pinAction)pinRobotCharacteristic[0];
```
Comment (lucarin91) [inline sketches/BLE_Scratch/BLE_Scratch.ino:520]: in cpp, you don't need to use write again `enum`.  I will also use the more idiomatic `static_cast`
```suggestion
  pinAction action = static_cast<pinAction>(pinRobotCharacteristic[0]);
```

**Review by lucarin91 — CHANGES_REQUESTED** [review]

File: scratch-arduino-extensions/scratch-vm/src/extensions/scratch3_arduino/peripheral.js [hunk 18]
```diff
@@ -475,6 +486,66 @@ class ArduinoPeripheral {
...
     }
+    moveForward(steps) {
+        const robotAction = [
+            ROBOT_ACTIONS.MOVE_FORWARD,
+            steps
+        ];
+        return this._send(BLECommand.ROBOT_ACTIONS, robotAction);
+    }
+    moveBackward(steps) {
+
```
Comment (lucarin91) [inline scratch-arduino-extensions/scratch-vm/src/extensions/scratch3_arduino/peripheral.js:497]: nit 
```suggestion
```

File: scratch-arduino-extensions/scratch-vm/src/extensions/scratch3_arduino/index.js [hunk 19]
```diff
@@ -199,55 +214,39 @@ class IncludeRobot {
             menus: {}
         };
     }
-
-    goForward(args) {
-        let speed = Cast.toNumber(args.SPEED);
-        speed = Math.max(0, Math.min(100, speed));
-
-        return this.setServoMotors({ SPEED_LEFT: -speed, SPEED_RIGHT: speed });
+    moveForward(args) {
```
Comment (lucarin91) [inline scratch-arduino-extensions/scratch-vm/src/extensions/scratch3_arduino/index.js:217]:
```suggestion

    moveForward(args) {
```

File: scratch-arduino-extensions/scratch-vm/src/extensions/scratch3_arduino/index.js [hunk 20]
```diff
@@ -199,55 +214,39 @@ class IncludeRobot {
...
-    goBackward(args) {
-        let speed = Cast.toNumber(args.SPEED);
-        speed = Math.max(0, Math.min(100, speed));
-
-        return this.setServoMotors({ SPEED_LEFT: speed, SPEED_RIGHT: -speed });
+    moveBackward(args) {
+        let steps = Cast.toNumber(args.STEPS);
+        return DEVICE.moveBackward(steps);
+    }
+    moveForwardTime(args) {
```
Comment (lucarin91) [inline scratch-arduino-extensions/scratch-vm/src/extensions/scratch3_arduino/index.js:225]:
```suggestion

    moveForwardTime(args) {
```

File: scratch-arduino-extensions/scratch-vm/src/extensions/scratch3_arduino/index.js [hunk 21]
```diff
@@ -199,55 +214,39 @@ class IncludeRobot {
...
-        let speed = Cast.toNumber(args.SPEED);
-        speed = Math.max(0, Math.min(100, speed));
-
-        return this.setServoMotors({ SPEED_LEFT: speed, SPEED_RIGHT: speed });
+        let seconds = Cast.toNumber(args.SECONDS);
+        let ms = Math.floor(seconds * 1000);
+        return DEVICE.turnLeft(ms)
     }
-
     turnRight(args) {
```
Comment (lucarin91) [inline scratch-arduino-extensions/scratch-vm/src/extensions/scratch3_arduino/index.js:240]:
```suggestion

    turnRight(args) {
```

File: scratch-arduino-extensions/scratch-vm/src/extensions/scratch3_arduino/peripheral.js [hunk 22]
```diff
@@ -475,6 +486,66 @@ class ArduinoPeripheral {
...
+    moveBackwardTime(milliseconds) {
+        let [arg0, arg1] = convertUInt16ToBytes(milliseconds);
+        const robotAction = [
+            ROBOT_ACTIONS.MOVE_BACKWARD_TIME,
+            arg0,
+            arg1
+        ];
+        return this._send(BLECommand.ROBOT_ACTIONS, robotAction);
+    }
+    turnLeft(milliseconds) {
```
Comment (lucarin91) [inline scratch-arduino-extensions/scratch-vm/src/extensions/scratch3_arduino/peripheral.js:523]:
```suggestion

    turnLeft(milliseconds) {
```

File: scratch-arduino-extensions/scratch-vm/src/extensions/scratch3_arduino/peripheral.js [hunk 23]
```diff
@@ -475,6 +486,66 @@ class ArduinoPeripheral {
...
+        let [arg0, arg1] = convertUInt16ToBytes(milliseconds);
+        const robotAction = [
+            ROBOT_ACTIONS.TURN_LEFT,
+            arg0,
+            arg1
+        ];
+        return this._send(BLECommand.ROBOT_ACTIONS, robotAction);
+    }
+    turnRight(milliseconds){
```
Comment (lucarin91) [inline scratch-arduino-extensions/scratch-vm/src/extensions/scratch3_arduino/peripheral.js:532]:
```suggestion

    turnRight(milliseconds){
```

**Review by lucarin91 — COMMENTED** [review]
add/remove some spaces 😄

**Pull request merged**